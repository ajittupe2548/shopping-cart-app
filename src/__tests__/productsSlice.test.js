import { configureStore } from "@reduxjs/toolkit";
import productsReducer, { fetchProducts } from "../store/productsSlice";

// Mock fetch
global.fetch = jest.fn();

describe("productsSlice", () => {
  let store;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        products: productsReducer,
      },
    });
    fetch.mockClear();
  });

  const mockProducts = [
    {
      id: 1,
      title: "Test Product 1",
      price: 29.99,
      description: "Test description",
      category: "test",
      image: "test1.jpg",
      rating: { rate: 4.5, count: 120 },
    },
    {
      id: 2,
      title: "Test Product 2",
      price: 39.99,
      description: "Test description 2",
      category: "test",
      image: "test2.jpg",
      rating: { rate: 4.0, count: 80 },
    },
  ];

  test("should handle initial state", () => {
    const state = store.getState().products;
    expect(state.items).toEqual([]);
    expect(state.status).toBe("idle");
    expect(state.error).toBe(null);
  });

  test("should handle fetchProducts.pending", () => {
    store.dispatch(fetchProducts.pending());
    const state = store.getState().products;

    expect(state.status).toBe("loading");
    expect(state.error).toBe(null);
  });

  test("should handle fetchProducts.fulfilled", () => {
    store.dispatch(fetchProducts.fulfilled(mockProducts));
    const state = store.getState().products;

    expect(state.status).toBe("succeeded");
    expect(state.items).toEqual(mockProducts);
    expect(state.error).toBe(null);
  });

  test("should handle fetchProducts.rejected", () => {
    const errorMessage = "Failed to fetch products";
    store.dispatch(fetchProducts.rejected(null, "", "", errorMessage));
    const state = store.getState().products;

    expect(state.status).toBe("failed");
    expect(state.error).toBe(errorMessage);
  });

  test("should fetch products successfully", async () => {
    fetch.mockResolvedValue({
      ok: true,
      json: async () => mockProducts,
    });

    await store.dispatch(fetchProducts());
    const state = store.getState().products;

    expect(state.status).toBe("succeeded");
    expect(state.items).toEqual(mockProducts);
    expect(fetch).toHaveBeenCalledWith("https://fakestoreapi.com/products");
  });

  test("should handle fetch products error", async () => {
    const errorMessage = "Network Error";
    fetch.mockRejectedValue(new Error(errorMessage));

    await store.dispatch(fetchProducts());
    const state = store.getState().products;

    expect(state.status).toBe("failed");
    expect(state.error).toBe(errorMessage);
  });

  test("should handle HTTP error status", async () => {
    fetch.mockResolvedValue({
      ok: false,
      status: 404,
    });

    await store.dispatch(fetchProducts());
    const state = store.getState().products;

    expect(state.status).toBe("failed");
    expect(state.error).toBe("HTTP error! status: 404");
  });

  test("matches snapshot with initial state", () => {
    const state = store.getState().products;
    expect(state).toMatchSnapshot();
  });

  test("matches snapshot with loaded products", async () => {
    fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => mockProducts,
    });

    await store.dispatch(fetchProducts());
    const state = store.getState().products;
    expect(state).toMatchSnapshot();
  });
});
