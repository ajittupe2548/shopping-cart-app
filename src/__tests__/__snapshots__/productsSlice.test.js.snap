// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`productsSlice matches snapshot with initial state 1`] = `
{
  "error": null,
  "items": [],
  "status": "idle",
}
`;

exports[`productsSlice matches snapshot with loaded products 1`] = `
{
  "error": null,
  "items": [
    {
      "category": "test",
      "description": "Test description",
      "id": 1,
      "image": "test1.jpg",
      "price": 29.99,
      "rating": {
        "count": 120,
        "rate": 4.5,
      },
      "title": "Test Product 1",
    },
    {
      "category": "test",
      "description": "Test description 2",
      "id": 2,
      "image": "test2.jpg",
      "price": 39.99,
      "rating": {
        "count": 80,
        "rate": 4,
      },
      "title": "Test Product 2",
    },
  ],
  "status": "succeeded",
}
`;
